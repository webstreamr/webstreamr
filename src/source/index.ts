export * from './CineHDPlus';
export * from './Cuevana';
export * from './Eurostreaming';
export * from './Frembed';
export * from './FrenchCloud';
export * from './KinoGer';
export * from './MegaKino';
export * from './MeineCloud';
export * from './MostraGuarda';
export * from './Movix';
export * from './PrimeWire';
export * from './StreamKiste';
export * from './VerHdLink';
export * from './VidSrc';
export * from './VixSrc';
export * from './types';
