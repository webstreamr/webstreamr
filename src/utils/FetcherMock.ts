/* istanbul ignore file */
import crypto from 'crypto';
import fs from 'node:fs';
import CachePolicy from 'http-cache-semantics';
import slugify from 'slugify';
import { RequestInit } from 'undici';
import winston from 'winston';
import { Context } from '../types';
import { envGet } from './env';
import { Fetcher } from './Fetcher';

export class FetcherMock extends Fetcher {
  private readonly fixturePath: string;

  public constructor(fixturePath: string) {
    super(winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] }));

    this.fixturePath = fixturePath;
  }

  public override async text(ctx: Context, url: URL, init?: RequestInit): Promise<string> {
    const path = `${this.fixturePath}/${this.slugifyUrl(url)}`;

    return this.fetch(path, ctx, url, init);
  };

  public override async textPost(ctx: Context, url: URL, body: string, init?: RequestInit): Promise<string> {
    const path = `${this.fixturePath}/post-${this.slugifyUrl(url)}-${slugify(body)}`;

    return this.fetch(path, ctx, url, { ...init, method: 'POST', body });
  };

  public override async head(ctx: Context, url: URL, init?: RequestInit): Promise<CachePolicy.Headers> {
    const path = `${this.fixturePath}/head-${this.slugifyUrl(url)}`;

    return JSON.parse(await this.fetch(path, ctx, url, { ...init, method: 'HEAD' }));
  };

  private readonly slugifyUrl = (url: URL): string => {
    if (url.href.length > 255) {
      return slugify(`${url.origin}-${crypto.createHash('md5').update(url.href).digest('hex')}`);
    }

    return slugify(url.href);
  };

  private readonly fetch = async (path: string, ctx: Context, url: URL, init?: RequestInit): Promise<string> => {
    const errorPath = `${path}.error`;

    if (fs.existsSync(errorPath)) {
      throw new Error(fs.readFileSync(errorPath).toString());
    } else if (fs.existsSync(path)) {
      return fs.readFileSync(path).toString();
    } else {
      let response;
      try {
        if (envGet('TEST_UPDATE_FIXTURES')) {
          response = await super.fetchWithTimeout(ctx, url, init);
        } else {
          console.error(`No fixture found at "${path}".`);
          process.exit(1);
        }
      } catch (error) {
        fs.writeFileSync(errorPath, `${error}`);
        throw error;
      }

      if (response.status < 200 || response.status > 399) {
        const message = `Fetcher error: ${response.status}: ${response.statusText}`;
        fs.writeFileSync(errorPath, message);
        throw new Error(message);
      }

      let result;
      if (init?.method === 'HEAD') {
        const headers: Record<string, string> = {};

        response.headers.forEach((value, key) => {
          headers[key] = value;
        });

        result = JSON.stringify(headers);
      } else {
        result = await response.text();
      }

      fs.writeFileSync(path, result);

      return result;
    }
  };
}
