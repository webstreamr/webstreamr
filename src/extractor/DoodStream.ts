import randomstring from 'randomstring';
import { Extractor } from './types';
import { Fetcher } from '../utils';
import { Context } from '../types';

export class DoodStream implements Extractor {
  readonly id = 'doodstream';

  readonly label = 'DoodStream';

  readonly ttl = 900000; // 15m

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (url: URL): boolean => null !== url.host.match(/dood/);

  readonly extract = async (ctx: Context, url: URL, countryCode: string) => {
    const videoId = url.pathname.split('/').slice(-1)[0] as string;
    const normalizedUrl = new URL(`http://dood.to/e/${videoId}`);

    const html = await this.fetcher.text(ctx, new URL(normalizedUrl));

    const passMd5Match = html.match(/\/pass_md5\/[\w-]+\/([\w-]+)/) as string[];

    const token = passMd5Match[1] as string;

    const baseUrl = await this.fetcher.text(ctx, new URL(`http://dood.to${passMd5Match[0]}`));

    return {
      url: new URL(`${baseUrl + randomstring.generate(10)}?token=${token}&expiry=${Date.now()}`),
      label: this.label,
      sourceId: `${this.id}_${countryCode.toLowerCase()}`,
      height: 0,
      bytes: 0,
      countryCode,
      requestHeaders: {
        Referer: 'http://dood.to/',
      },
    };
  };
}
