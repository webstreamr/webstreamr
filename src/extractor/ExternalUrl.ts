import { Extractor } from './types';
import { Fetcher } from '../utils';
import { Context, Meta } from '../types';

export class ExternalUrl implements Extractor {
  readonly id = 'external';

  readonly label = 'External';

  readonly ttl = 3600000; // 1h

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (ctx: Context, url: URL): boolean => !('excludeExternalUrls' in ctx.config) && null !== url.host.match(/.*/);

  readonly extract = async (ctx: Context, url: URL, meta: Meta) => {
    // We only want to make sure that the URL is accessible
    await this.fetcher.head(ctx, url);

    return {
      url: url,
      isExternal: true,
      label: `${url.host}`,
      sourceId: `${this.id}_${meta.countryCode.toLowerCase()}`,
      ttl: this.ttl,
      meta,
    };
  };
}
