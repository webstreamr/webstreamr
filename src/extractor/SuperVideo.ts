import bytes from 'bytes';
import * as cheerio from 'cheerio';
import { Extractor } from './types';
import { extractUrlFromPacked, Fetcher } from '../utils';
import { Context, Meta } from '../types';

export class SuperVideo implements Extractor {
  readonly id = 'supervideo';

  readonly label = 'SuperVideo';

  readonly ttl = 900000; // 15m

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (_ctx: Context, url: URL): boolean => null !== url.host.match(/supervideo/);

  readonly extract = async (ctx: Context, url: URL, meta: Meta) => {
    url.pathname = url.pathname.replace('/e/', '').replace('/embed-', '/');
    const html = await this.fetcher.text(ctx, url);

    const heightAndSizeMatch = html.match(/\d{3,}x(\d{3,}), ([\d.]+ ?[GM]B)/) as string[];

    const $ = cheerio.load(html);
    const title = $('title').text().trim().replace(/^Watch /, '').trim();

    return {
      url: extractUrlFromPacked(html, [/sources:\[{file:"(.*?)"/]),
      label: this.label,
      sourceId: `${this.id}_${meta.countryCode.toLowerCase()}`,
      ttl: this.ttl,
      meta: {
        bytes: bytes.parse(heightAndSizeMatch[2] as string) as number,
        height: parseInt(heightAndSizeMatch[1] as string) as number,
        title,
        ...meta,
      },
    };
  };
}
