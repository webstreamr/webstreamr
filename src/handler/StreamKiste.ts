import { ContentType } from 'stremio-addon-sdk';
import * as cheerio from 'cheerio';
import { Handler } from './types';
import { ImdbId, Fetcher, getImdbId, Id } from '../utils';
import { ExtractorRegistry } from '../extractor';
import { Context, CountryCode } from '../types';

export class StreamKiste implements Handler {
  readonly id = 'streamkiste';

  readonly label = 'StreamKiste';

  readonly contentTypes: ContentType[] = ['series'];

  readonly countryCodes: CountryCode[] = ['de'];

  private readonly fetcher: Fetcher;
  private readonly extractorRegistry: ExtractorRegistry;

  constructor(fetcher: Fetcher, extractorRegistry: ExtractorRegistry) {
    this.fetcher = fetcher;
    this.extractorRegistry = extractorRegistry;
  }

  readonly handle = async (ctx: Context, _type: string, id: Id) => {
    const imdbId = await getImdbId(ctx, this.fetcher, id);

    const seriesPageUrl = await this.fetchSeriesPageUrl(ctx, imdbId);
    if (!seriesPageUrl) {
      return [];
    }

    const html = await this.fetcher.text(ctx, seriesPageUrl);

    const $ = cheerio.load(html);

    const title = $('meta[property="og:title"]').attr('content') as string;

    return Promise.all(
      $(`[data-num="${imdbId.season}x${imdbId.episode}"]`)
        .siblings('.mirrors')
        .children('[data-link]')
        .map((_i, el) => new URL(($(el).attr('data-link') as string).replace(/^(https:)?\/\//, 'https://')))
        .toArray()
        .filter(url => !url.host.match(/streamkiste/))
        .map(url => this.extractorRegistry.handle({ ...ctx, referer: seriesPageUrl }, url, { countryCode: 'de', title: `${title.trim()} ${imdbId.season}x${imdbId.episode}` })),
    );
  };

  private fetchSeriesPageUrl = async (ctx: Context, imdbId: ImdbId): Promise<URL | undefined> => {
    const html = await this.fetcher.text(ctx, new URL(`https://streamkiste.taxi/?story=${imdbId.id}&do=search&subaction=search`));

    const $ = cheerio.load(html);

    const url = $('.res_item a[href]:first')
      .map((_i, el) => $(el).attr('href'))
      .get(0);

    return url !== undefined ? new URL(url) : url;
  };
}
