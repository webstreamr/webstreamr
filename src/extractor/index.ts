import { Fetcher } from '../utils';
import { DoodStream } from './DoodStream';
import { Dropload } from './Dropload';
import { ExternalUrl } from './ExternalUrl';
import { Extractor } from './Extractor';
import { Fastream } from './Fastream';
import { FileLions } from './FileLions';
import { HubCloud } from './HubCloud';
import { KinoGer } from './KinoGer';
import { Mixdrop } from './Mixdrop';
import { SaveFiles } from './SaveFiles';
import { Soaper } from './Soaper';
import { StreamEmbed } from './StreamEmbed';
import { Streamtape } from './Streamtape';
import { SuperVideo } from './SuperVideo';
import { Uqload } from './Uqload';
import { VidSrc } from './VidSrc';
import { VixSrc } from './VixSrc';
import { XPrime } from './XPrime';
import { YouTube } from './YouTube';

export * from './Extractor';
export * from './ExtractorRegistry';

export const createExtractors = (fetcher: Fetcher): Extractor[] => [
  new DoodStream(fetcher),
  new Dropload(fetcher),
  new Fastream(fetcher),
  new FileLions(fetcher),
  new HubCloud(fetcher),
  new KinoGer(fetcher),
  new Mixdrop(fetcher),
  new SaveFiles(fetcher),
  new Soaper(fetcher),
  new StreamEmbed(fetcher),
  new Streamtape(fetcher),
  new SuperVideo(fetcher),
  new Uqload(fetcher),
  new VidSrc(fetcher, ['in', 'net', 'xyz', 'io', 'vc']), // https://vidsrc.domains/
  new VixSrc(fetcher),
  new XPrime(fetcher),
  new YouTube(fetcher),
  new ExternalUrl(fetcher), // fallback extractor which must come last
];
