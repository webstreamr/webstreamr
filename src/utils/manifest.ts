import { flag } from 'country-emoji';
import { Handler } from '../handler';
import { Config, ManifestWithConfig } from '../types';

export const buildManifest = (handlers: Handler[], config: Config): ManifestWithConfig => {
  const manifest: ManifestWithConfig = {
    id: process.env['MANIFEST_ID'] || 'webstreamr',
    version: '0.21.0', // x-release-please-version
    name: process.env['MANIFEST_NAME'] || 'WebStreamr',
    description: 'Provides HTTP URLs from streaming websites.',
    resources: [
      'stream',
    ],
    types: [
      'movie',
      'series',
    ],
    catalogs: [],
    idPrefixes: ['tt'],
    behaviorHints: {
      p2p: false,
      configurable: true,
      configurationRequired: Object.keys(config).length === 0,
    },
    config: [],
    // @ts-expect-error inofficial prop needed for add-on claiming on https://stremio-addons.net
    stremioAddonsConfig: {
      issuer: 'https://stremio-addons.net',
      signature: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMTI4Q0JDLUhTMjU2In0..D3-Wl15vxnYltwx8G52rRg.16-0VzsSgTvnv0wampIR8YPZsFhH8-7IgpuqxcUfC2p7kIJtuk8xQzrzqQOLXANEpaH_w7a4JOEpNo8wv28Zo_nMGCLOXgWbyGM3sQ-tfq6DCK3JU0ol6YMC1UDX2z_c.IOLlZyTYx_swutjYSTES0Q',
    },
  };

  const countryCodeHandlers: Record<string, Handler[]> = {};
  handlers.forEach((handler) => {
    handler.countryCodes.forEach(countryCode => countryCodeHandlers[countryCode] = [...(countryCodeHandlers[countryCode] ?? []), handler]);
  });

  const sortedLanguageHandlers = Object.entries(countryCodeHandlers)
    .sort(([countryCodeA], [countryCodeB]) => countryCodeA.localeCompare(countryCodeB));

  for (const [countryCode, handlers] of sortedLanguageHandlers) {
    manifest.config.push({
      key: countryCode,
      type: 'checkbox',
      title: `${flag(countryCode)} (${handlers.map(handler => handler.label).join(', ')})`,
      ...(countryCode in config && { default: 'checked' }),
    });
  }

  return manifest;
};
