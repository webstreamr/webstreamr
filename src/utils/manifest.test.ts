import { DoodStream } from '../extractor/DoodStream';
import { ExternalUrl } from '../extractor/ExternalUrl';
import { SuperVideo } from '../extractor/SuperVideo';
import { createSources } from '../source';
import { MeineCloud } from '../source/MeineCloud';
import { StreamKiste } from '../source/StreamKiste';
import { VerHdLink } from '../source/VerHdLink';
import { VixSrc } from '../source/VixSrc';
import { FetcherMock } from './FetcherMock';
import { buildManifest } from './manifest';

const fetcher = new FetcherMock('/dev/null');

describe('buildManifest', () => {
  test('default manifest', async () => {
    const manifest = buildManifest(createSources(fetcher), [], {});

    expect(manifest).toMatchSnapshot({
      version: expect.any(String),
    });
  });

  test('has unchecked source without a config', () => {
    const sources = [
      new VixSrc(fetcher),
      new VerHdLink(fetcher),
      new StreamKiste(fetcher),
      new MeineCloud(fetcher),
    ];

    const manifest = buildManifest(sources, [], {});

    expect(manifest.config).toMatchSnapshot();
  });

  test('has checked source with appropriate config', () => {
    const sources = [
      new VerHdLink(fetcher),
      new StreamKiste(fetcher),
      new MeineCloud(fetcher),
    ];
    const manifest = buildManifest(sources, [], { de: 'on', includeExternalUrls: 'on' });

    expect(manifest.config).toMatchSnapshot();
  });

  test('showErrors and includeExternalUrls are unchecked by default', () => {
    const manifest = buildManifest([], [], {});

    expect(manifest.config).toMatchSnapshot();
  });

  test('has checked showErrors and includeExternalUrls', () => {
    const manifest = buildManifest([], [], { showErrors: 'on', includeExternalUrls: 'on' });

    expect(manifest.config).toMatchSnapshot();
  });

  test('disable extractors', () => {
    const extractors = [
      new DoodStream(fetcher),
      new SuperVideo(fetcher),
      new ExternalUrl(fetcher),
    ];
    const manifest = buildManifest([], extractors, { disableExtractor_doodstream: 'on' });

    expect(manifest.config).toMatchSnapshot();
  });
});
