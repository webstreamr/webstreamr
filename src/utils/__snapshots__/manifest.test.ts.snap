// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`buildManifest has checked includeExternalUrls 1`] = `
[
  {
    "default": "checked",
    "key": "includeExternalUrls",
    "title": "Include external URLs in results",
    "type": "checkbox",
  },
  {
    "default": "",
    "key": "mediaFlowProxyUrl",
    "title": "MediaFlow Proxy URL",
    "type": "text",
  },
  {
    "default": "",
    "key": "mediaFlowProxyPassword",
    "title": "MediaFlow Proxy Password",
    "type": "password",
  },
]
`;

exports[`buildManifest has checked source with appropriate config 1`] = `
[
  {
    "default": "checked",
    "key": "de",
    "title": "German ðŸ‡©ðŸ‡ª (StreamKiste, MeineCloud)",
    "type": "checkbox",
  },
  {
    "key": "es",
    "title": "Castilian Spanish ðŸ‡ªðŸ‡¸ (VerHdLink)",
    "type": "checkbox",
  },
  {
    "key": "mx",
    "title": "Latin American Spanish ðŸ‡²ðŸ‡½ (VerHdLink)",
    "type": "checkbox",
  },
  {
    "default": "checked",
    "key": "includeExternalUrls",
    "title": "Include external URLs in results",
    "type": "checkbox",
  },
  {
    "default": "",
    "key": "mediaFlowProxyUrl",
    "title": "MediaFlow Proxy URL",
    "type": "text",
  },
  {
    "default": "",
    "key": "mediaFlowProxyPassword",
    "title": "MediaFlow Proxy Password",
    "type": "password",
  },
]
`;

exports[`buildManifest has unchecked source without a config 1`] = `
[
  {
    "key": "de",
    "title": "German ðŸ‡©ðŸ‡ª (StreamKiste, MeineCloud)",
    "type": "checkbox",
  },
  {
    "key": "es",
    "title": "Castilian Spanish ðŸ‡ªðŸ‡¸ (VerHdLink)",
    "type": "checkbox",
  },
  {
    "key": "mx",
    "title": "Latin American Spanish ðŸ‡²ðŸ‡½ (VerHdLink)",
    "type": "checkbox",
  },
  {
    "key": "includeExternalUrls",
    "title": "Include external URLs in results",
    "type": "checkbox",
  },
  {
    "default": "",
    "key": "mediaFlowProxyUrl",
    "title": "MediaFlow Proxy URL",
    "type": "text",
  },
  {
    "default": "",
    "key": "mediaFlowProxyPassword",
    "title": "MediaFlow Proxy Password",
    "type": "password",
  },
]
`;

exports[`buildManifest includeExternalUrls is unchecked by default 1`] = `
[
  {
    "key": "includeExternalUrls",
    "title": "Include external URLs in results",
    "type": "checkbox",
  },
  {
    "default": "",
    "key": "mediaFlowProxyUrl",
    "title": "MediaFlow Proxy URL",
    "type": "text",
  },
  {
    "default": "",
    "key": "mediaFlowProxyPassword",
    "title": "MediaFlow Proxy Password",
    "type": "password",
  },
]
`;
